src/System.Text.Json/docs/writable_json_dom_spec.md

```diff
@@ -0,0 +1,257 @@
+
+* Implicit operators for `JsonString`, `JsonBoolean` and `JsonNumber` as an additional feature.
+* `Sort` not implemented for `JsonArray`, beacuse there is no right way to compare `JsonObject`s. If user wants to sort `JsonArray` of `JsonNumber`s, `JsonBooleans`s or `JsonStrings` he/she now needs to do the following: convert `JsonArray` to regular array (by iterating through all elements), calling sort (and converting back to `JsonArray` if needed).
+* No support for duplicates of property names. Possibly, adding an option for user to choose from: "first value", "last value", or throw-on-duplicate.
+* Transformation API:
+    * `DeepCopy` method in JsonElement allowing to change JsonElement and JsonDocument into JsonNode recursively transforming all of the elements
+    * `AsJsonElement` method in JsonNode allowing to change JsonNode into JsonElement with IsImmutable property set to false
+    * `IsImmutable` property informing if JsonElement is keeping JsonDocument or JsonNode underneath
+    * `Parse(string)` in JsonNode to be able to parse Json string right into JsonNode if user knows he/she wants mutable version
+    * `DeepCopy` in JsonNode to make a copy of the whole tree
+    * `GetNode` and TryGetNode in JsonNode allowing to retrieve it from JsonElement
+
+## Implementation details
+* `JsonNumber` value is stored as a `string`.
+
+## Open questions
+API:
+* Do we want to add recursive equals on `JsonArray` and `JsonObject`?
+* Do we want to make `JsonNode`s derived types (and which) implement `IComparable`?
+* Would escaped characters be supported for creating `JsonNumber` from string? 
+* Is the API for `JsonNode` and `JsonElement` interactions sufficient? 
+* Do we want to support duplicate and order preservation/control when adding/removing values in `JsonArray`/`JsonObject`?
+* Do we want to have accelerator APIs for writing a JsonNode to a Utf8JsonWriter without having to go through JsonElement? (currently added as a `Parse` method in JsonNode)
+* Should JsonNode have a JsonValueKind property that a caller can inspect and cast to the right concrete type?
+* Let's say someone else passes a JsonNode to me, what can I do with it?
+* Should nodes track their own position in the JSON graph? Do we want to allow properties like Parent, Next and Previous?
+
+    | Solution | Pros | Cons |
+    |----------|:-------------|--------|
+    |current API| - no additional checks need to be made | - creating recursive loop by the user may be problematic |
+    |tracking nodes | - handles recursive loop problem | - when node is added to a parent, it needs to be checked <br>  if it already has a parent  and make a copy if it has |
+
+Implementation:
+* Do we want to add a copy of `JsonWriterHelper.ValidateNumber` with additional checks?
+* Do we want to change `JsonNumber`'s backing field to something different than `string`?     
+    Suggestions: `Span<byte>` or array of `Utf8String`/`Char8` (once they come online in the future) / `byte`  
```

Comment:
Instead of string or bytes, what about internal types that inherit from JsonNumber:

Int32JsonNumber, SingleJsonNumber, DoubleJsonNumber, DecimalJsonNumber, BigIntegerJsonNumber, etc

These would hold the precise .NET type - no boxing or allocations of a string - and wouldn't loose precision. JsonNumber would have an internal ctor and instances would be created via a factory method: `JsonNumber.Create(1000000000.99m)`. This factory method would be used when doing implicit conversion to `JsonNumber`:

```cs
JsonNode integer = 1;
JsonNode dec = 1.0001m;
JsonNode bigInteger = new BigInteger(10000000000);
```

Reply:
That's an interesting approach. Would we then remove the capability of creating a JsonNumber from a string (or text of any encoding)? If not, in your approach would we infer the best fit concrete numeric type to map the string to?

Reply:
My interpretation on James' suggestion is that the build-from-string version would just have the current implementation.

The internal UIntJsonNumber and such would have to be held in a field of the JsonNumber, though, so that the SetValue methods allow it to change datatype.  E.g.

```C#
JsonArray array = (JsonArray)JsonNode.Parse("[1, 2, 3]");
array[2].SetValue(1.21m);
array.WriteTo(someWriterIHadDontAsk);
```

(Modulo real method names).

Reply:
> The internal UIntJsonNumber and such would have to be held in a field of the JsonNumber

That wasn't what I meant. That would basically just be manual boxing.

I was thinking something like:

```cs
public abstract class JsonNumber
{
    public JsonTokenType Type => JsonNodeType.Number;

    public static JsonNumber Create(int i) => return new Int32JsonNumber(i);
    public static JsonNumber Create(decimal d) => return new DecimalJsonNumber(d);

    public abstract int GetInteger();
    public abstract decimal GetDecimal();
}

internal abstract class Int32JsonNumber : JsonNumber
{
    private readonly int _value;
    internal Int32JsonNumber(int value)
    {
        _value = value;
    }

    public override int GetInteger() => _value;
    public override int GetDecimal() => _value;
}
```

JsonNumber would need to be immutable for this to work.

Reply:
Yeah, sorry, I combined two things.  1) JsonNode.Create(string) (so that you could transmit `1e400`, if so desired) would be the current implementation.

2) That, given that we've said the value should be mutable, it would need to be a deferred field.

If we think that the "primitive" values don't need to be mutable, (just the object and arrray ones) then directly returning specialized values would work.

Reply:
Json.NET's JValue can be modified. I'm not sure how valuable that is.

Making JsonString/JsonBoolean/JsonNumber immutable is something to explore.

Reply:
Should `Int32JsonNumber` be `internal sealed`?

Reply:
Thank you very much for the suggestion, I will add it here! I think it would be best to keep the current implementation for now and come back to the discussion during the API review. I will concentrate on trying to implement other pieces of API as my internship time is limited, I can also prepare the comparison of these solutions later on.

Reply:
That sounds great to me!